v2api/misfiled_ammendments_dedupe.py

```diff
@@ -0,0 +1,99 @@
+import pandas as pd
+import datetime as dt
+import numpy as np
+import requests
+from .query_v2_api import AUTH, BASE_URL, PARAMS
+
+def get_form(form,offset=0):
+    """ Get filings with matching form type
+    """
+    url = f'{BASE_URL}/filing/v101/filings?Limit=100000&SpecificationForm={form}'
+
+    params = { **PARAMS }
+    if offset > 0:
+        params['offset'] = offset
+
+    res = requests.get(url, params=params, auth=AUTH)
+    if res.status_code == 500:
+        return get_form(form,offset=0)
+    else:
+        body = res.json()
+        results = body.pop('results')
+
+        return results, body
+    
+class form460:
+    def __init__(self, rpt_date, date_range, filing_id) -> None:
+        self.date_range = date_range
+        self.rpt_date = rpt_date
+        self.filing_id = filing_id        
+    def update(self, new_date_range):
+        self.date_range = new_date_range
+
+def is_iso_str_in_range(iso_str, date_range):
+    if iso_str:
+        return dt.datetime.fromisoformat(iso_str).date().toordinal() in date_range
+    else:
+        return True
+
+# This should cut the complexity from O(n^2) to O(n)
+def date_range_fix(df, result_dict):    
+    array1 = np.array(df['filingNid'])
+    array2 = np.array(df['date_range'].apply(lambda x: x.date_range))
+    i = 0
+    for array in array2[1:]:    
+        if array2[i]:
+            array2[i+1:] = array2[i+1:] - np.repeat(array2[i], len(array2[i+1:]))
+        i+=1
+    result_dict.update(dict(zip(array1, array2)))
+    return result_dict
```

Comment:
The data has been ranked by report date by sorting the DataFrame, we create a vector of the date_ranges (periods covered) and the filing ids. Then we iterate through the vector of date_ranges starting at the second date_range so that the operation is performed for each date_range with a date_range bellow it. We subtract the date_ranges of forms from all forms with a lower priority in 1 operation per form by creating a vector with the high priority form with length of all lower priority forms.
Then the resulting array is zipped back together with the filling ids giving the dictionary to compare transactions to.

Reply:
love this 😍